MeshToLevelSet : Support Merging Inputs

danieldresser-ie opened the pull request:
Ended up requiring a few miscellany pieces of groundwork, but the actual changes to MeshToLevelSet are pretty straightforward.

The commit: "PrimitiveAlgo::mergePrimitives : Optimize case of single input" is perhaps worth mentioning ... it's not strictly needed here, but I was thinking of adding special case to MeshToLevelSet to make sure we didn't sacrifice perf in the single input case, and figured it made more sense to put that special case in mergePrimitives instead, so we could keep MeshToLevelSet simpler, and benefit all clients of mergePrimitives.

johnhaddon commented on src/GafferVDB/MeshToLevelSet.cpp:
```diff
@@ -184,66 +188,102 @@ const FloatPlug *MeshToLevelSet::interiorBandwidthPlug() const
 	;
 }
 
-void MeshToLevelSet::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+void MeshToLevelSet::hashMergedObject(
+	const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h
+) const
 {
-	ObjectProcessor::hashProcessedObject( path, context, h );
+	MergeObjects::hashMergedObject( path, context, h );
```
It's unfortunate that `MergeObjects::hashMergedObject()` includes the hash of the input transform. This means that in the simple case of converting a single mesh with the default destination, we now repeat the compute every time the transform changes. This is quite a noticeable performance difference, so I wonder if we could detect this case in the base class, and omit the transform from the hash?

johnhaddon commented on src/GafferVDB/MeshToLevelSet.cpp:
```diff
@@ -184,66 +188,102 @@ const FloatPlug *MeshToLevelSet::interiorBandwidthPlug() const
+IECore::ConstObjectPtr MeshToLevelSet::computeMergedObject( const std::vector< std::pair< IECore::ConstObjectPtr, Imath::M44f > > &sources, const Gaffer::Context *context ) const
 {
-	const MeshPrimitive *mesh = runTimeCast<const MeshPrimitive>( inputObject );
-	if( !mesh )
```
This is a change in behaviour that I hadn't anticipated. In the previous version, non-meshes were passed through, but they are now replaced with an empty level set. I think this is OK (unless anyone else has concerns?) but will need calling out in Changes.md.

johnhaddon reviewed (commented):
Thanks Daniel - great to see how straighforward the new base class has made this. In addition to the inline comments, we need a new Improvements entry in Changes.md. I think we could also benefit from a more specific description in the metadata for `destination`. This could talk about the "where the level set is placed" rather than "where locations are merged to", and also start by mentioning that by default that they are created in place, but can be merged by entering a different destination.

And just to prove that I've tested this a bit, I give you the ALab, after being passed through MeshToLevelSet and back through LevelSetToMesh : 

![image](https://github.com/user-attachments/assets/ff87f68e-40c3-4f27-9ff1-c7c4c10f1fc7)

danieldresser-ie commented:
Added [0b4a930](https://github.com/GafferHQ/gaffer/pull/6060/commits/0b4a930d59315e68f5c100e7fdc19b0df46b2e2c) to address the hashing concern, and updated docs and Changes.md

Should be mergeable now, though maybe we want to wait until I knock out LevelSetToMesh as well ( which should be straightforward ... I'm assuming from a correctness point of view, we want to union the level sets first, and then convert, so that you don't get internal faces ... we don't need to test if it would actually be faster to convert then merge, because if a user wants that, they would use a separate MergeMeshes? )

danieldresser-ie commented:
Added two more commits adding the same sort of support to LevelSetToMesh.

We still have an ongoing conversation about resampleToMatch vs doResampleToMatch ( both offline and discussed in the code comments ), but I'm now feeling a lot more confident that in any situation where there is a noticeable difference, doResampleToMatch yields a more correct result in this case.

Main thing I haven't addressed is that LevelSetToMesh now takes a pretty ridiculous amount of time to compile. It was already taking 1 minute before these changes, and is now taking 2 minutes. I actually tried a bit of refactoring to see if I could reduce this, but without any luck.

johnhaddon reviewed (commented):
Couple of small comments from me after one more look. There's also the question of the failing Windows CI - looks like it doesn't like compiling the new code - maybe @ericmehl might have some input?

ericmehl commented:
I can reproduce that on my workstation. It looks like we just need to enable `/bigobj` in the build flags on Windows for all configurations. This is already done on the debug configs for the same reason.

Based on discussion at https://developercommunity.visualstudio.com/t/Enable-bigobj-by-default/1031214, it sounds like the only reason to not enable it by default is for compatibility with pre-2005 era MSVC linkers.

As much as I like backwards compatibility, I think we're safe not being able to run Gaffer on MS-DOS.

danieldresser-ie commented:
Added more comments with some of the things we've discussed ( Why use doResampleToMatch, why not use QuadraticSampler, why not include the transform in the voxel size heuristic ).

Restricted compilation to just FloatGrid for now. That was sufficient to get it building on Windows, and knocked 60 seconds off the compile time.

Updated Changes to mention dropping support for other grid types for now.

johnhaddon commented:
Rebased to sort out Changes.md (we've released a2 since opening this). Merging.
